blockchain/blockchain.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2017 Monax Industries Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package blockchain
+
+import (
+	"time"
+
+	"sync"
+
+	acm "github.com/hyperledger/burrow/account"
+	"github.com/hyperledger/burrow/genesis"
+)
+
+// Immutable Root of blockchain
+type Root interface {
+	// ChainID precomputed from GenesisDoc
+	ChainID() string
+	// GenesisHash precomputed from GenesisDoc
+	GenesisHash() []byte
+	GenesisDoc() genesis.GenesisDoc
+}
+
+// Immutable pointer to the current tip of the blockchain
+type Tip interface {
```

Comment:
I like Head or ChainHead better, as I tend to think of a blockchain as a kind of linked list. Not a big deal.